materialize-snowflake/snowflake.go

```diff
@@ -552,6 +611,108 @@ func (d *transactor) Acknowledge(ctx context.Context) (*pf.ConnectorState, error
 		d.deleteFiles(ctx, []string{item.StagedDir})
 	}
 
+	// Keep asking for a report on the files that have been submitted for processing
+	// until they have all been successful, or an error has been thrown
+	var maxTries = 10
+	for pipeName, pipe := range pipes {
+		for tries := 0; tries < maxTries; tries++ {
+			// We first try to check the status of pipes using the REST API's insertReport
+			// The REST API does not wake up the warehouse, hence our preference
+			// however this API is not the most ergonomic and sometimes does not yield
+			// results as expected
+			report, err := d.pipeClient.InsertReport(pipeName, d.pipeMarkers[pipeName])
+			if err != nil {
+				return nil, fmt.Errorf("snowpipe: insertReports: %w", err)
+			}
+
+			if !report.CompleteResult {
+				tries--
+				d.pipeMarkers[pipeName] = ""
+			} else {
+				d.pipeMarkers[pipeName] = report.NextBeginMark
+			}
+
+			// If there are items still in progress, we continue retrying until those items
+			// resolve to another status
+			var hasItemsInProgress = false
+
+			// If the files have already been loaded, when we submit a request to
+			// load those files again, our request will not show up in reports.
+			// One way to find out whether the files were successfully loaded is to check
+			// the COPY_HISTORY to make sure they are there. The COPY_HISTORY is much more
+			// reliable. If they are not there, then something is wrong.
+			if len(report.Files) == 0 {
```

Comment:
Is there a scenario where the REST API gives us a partial list of files because some of the files have timed out of its retention, but some haven't? Is that covered by `report.CompleteResult`? I think we'd want to try `COPY_HISTORY` in that case.

Reply:
In that scenario eventually all files will time out and go out of retention period of REST API (10 minutes maximum), and then will ask `COPY_HISTORY`. It is very likely that files that we are checking as part of the same transaction will all time out close to each other, so if some of them have timed out, all of them are expected to time out soon.

Reply:
Also note that, once we have seen the partial list of files, we will persist the `beginMarker`, which means on the next run of `insertReport`, we will not get the same partial list of files. We will get a list of no values if the other files have expired, so that expediates our falling back to `COPY_HISTORY`.